src/jdk.compiler/share/classes/com/sun/tools/javac/tree/JCTree.java

```diff
@@ -623,6 +623,30 @@ public <R,D> R accept(TreeVisitor<R,D> v, D d) {
         public Tag getTag() {
             return TOPLEVEL;
         }
+
+        public boolean isAnonymousMainClass() {
+            // Detect anonymous main class.
+            for (JCTree def : defs) {
+                if (def.hasTag(Tag.METHODDEF) || def.hasTag(Tag.VARDEF)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public JCClassDecl getAnonymousMainClass() {
+            // Find anonymous main class.
+            for (JCTree def : defs) {
+                if (def.hasTag(CLASSDEF)) {
```

Comment:
Nit, in cases like this, we lately tend to write `def instanceof JCClassDecl cls`, although we understand this way to check the AST is only safe inside javac.

Reply:
Changing